File: backend/pkg/api/taskrun/handlers.go
package taskrun

import (
	"time"

	"backend/pkg/api/dto"
	"backend/pkg/api/middleware"
	"backend/pkg/api/response"
	"backend/pkg/core"
	"backend/pkg/core/taskrun"
	"backend/pkg/database"
	"backend/pkg/database/repository"
	"backend/pkg/types"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
)

func newTaskRunService() *taskrun.Service {
	return taskrun.NewService(repository.NewTaskRunRepository(database.DB()))
}

func toTaskRunResponse(run types.TaskRun) dto.TaskRunResponse {
	resp := dto.TaskRunResponse{
		ID:            run.ID.String(),
		Kind:          string(run.Kind),
		TaskName:      run.TaskName,
		Status:        string(run.Status),
		TriggerSource: run.TriggerSource,
		Summary:       run.Summary,
		ErrorSummary:  run.ErrorSummary,
		OutputSummary: run.OutputSummary,
		Metrics:       run.Metrics,
	}
	if run.ParentRunID != nil {
		parentID := run.ParentRunID.String()
		resp.ParentRunID = &parentID
	}
	if run.StartedAt != nil {
		value := run.StartedAt.Format(time.RFC3339)
		resp.StartedAt = &value
	}
	if run.CompletedAt != nil {
		value := run.CompletedAt.Format(time.RFC3339)
		resp.CompletedAt = &value
	}
	if run.StartedAt != nil && run.CompletedAt != nil {
		duration := run.CompletedAt.Sub(*run.StartedAt).Milliseconds()
		resp.DurationMS = &duration
	}
	return resp
}

func toTaskRunStepResponse(step types.TaskRunStep) dto.TaskRunStepResponse {
	resp := dto.TaskRunStepResponse{
		ID:           step.ID.String(),
		StepKey:      step.StepKey,
		StepName:     step.StepName,
		Status:       string(step.Status),
		Summary:      step.Summary,
		ErrorSummary: step.ErrorSummary,
		Metrics:      step.Metrics,
	}
	if step.StartedAt != nil {
		value := step.StartedAt.Format(time.RFC3339)
		resp.StartedAt = &value
	}
	if step.CompletedAt != nil {
		value := step.CompletedAt.Format(time.RFC3339)
		resp.CompletedAt = &value
	}
	return resp
}

func toTaskRunEventResponse(event types.TaskRunEvent, stepKey *string) dto.TaskRunEventResponse {
	return dto.TaskRunEventResponse{
		ID:        event.ID.String(),
		EventType: event.EventType,
		Level:     string(event.Level),
		Message:   event.Message,
		CreatedAt: event.CreatedAt.Format(time.RFC3339),
		StepKey:   stepKey,
		MetaData:  event.MetaData,
	}
}

func runFilterFromRequest(c *fiber.Ctx) (repository.TaskRunFilter, error) {
	userID, err := middleware.GetUserID(c)
	if err != nil {
		return repository.TaskRunFilter{}, core.UnauthorizedError("User not found")
	}
	filter := repository.TaskRunFilter{
		UserID:   &userID,
		TaskName: c.Query("task_name"),
		Status:   c.Query("status"),
		Kind:     c.Query("kind"),
		Limit:    c.QueryInt("limit", 50),
	}
	orgID := middleware.GetOrgID(c)
	if orgID != nil {
		filter.OrganizationID = orgID
		filter.UserID = nil
	}
	return filter, nil
}

func ensureRunAccess(c *fiber.Ctx, run *types.TaskRun) error {
	userID, err := middleware.GetUserID(c)
	if err != nil {
		return core.UnauthorizedError("User not found")
	}
	orgID := middleware.GetOrgID(c)

	if orgID != nil {
		if run.OrganizationID == nil || *run.OrganizationID != *orgID {
			return core.NotFoundError("Task run not found")
		}
		return nil
	}

	if run.UserID == nil || *run.UserID != userID {
		return core.NotFoundError("Task run not found")
	}
	return nil
}

func ListTaskRuns(c *fiber.Ctx) error {
	filter, err := runFilterFromRequest(c)
	if err != nil {
		return response.Error(c, err)
	}

	runs, err := newTaskRunService().ListRuns(c.Context(), filter)
	if err != nil {
		return response.Error(c, err)
	}

	items := make([]dto.TaskRunResponse, len(runs))
	for i, run := range runs {
		items[i] = toTaskRunResponse(run)
	}

	return response.Success(c, dto.TaskRunListResponse{Runs: items})
}

func GetTaskRun(c *fiber.Ctx) error {
	runID, err := uuid.Parse(c.Params("id"))
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid task run ID"))
	}

	service := newTaskRunService()
	run, err := service.GetRun(c.Context(), runID)
	if err != nil {
		return response.Error(c, err)
	}
	if err := ensureRunAccess(c, run); err != nil {
		return response.Error(c, err)
	}

	steps, err := service.ListStepsByRunID(c.Context(), runID)
	if err != nil {
		return response.Error(c, err)
	}
	stepKeysByID := map[uuid.UUID]string{}
	stepResponses := make([]dto.TaskRunStepResponse, len(steps))
	for i, step := range steps {
		stepKeysByID[step.ID] = step.StepKey
		stepResponses[i] = toTaskRunStepResponse(step)
	}

	events, err := service.ListEventsByRunID(c.Context(), runID)
	if err != nil {
		return response.Error(c, err)
	}
	eventResponses := make([]dto.TaskRunEventResponse, len(events))
	for i, event := range events {
		var stepKey *string
		if event.TaskRunStepID != nil {
			if key, ok := stepKeysByID[*event.TaskRunStepID]; ok {
				stepKey = &key
			}
		}
		eventResponses[i] = toTaskRunEventResponse(event, stepKey)
	}

	return response.Success(c, dto.TaskRunDetailResponse{
		Run:    toTaskRunResponse(*run),
		Steps:  stepResponses,
		Events: eventResponses,
	})
}

func ListTaskRunEvents(c *fiber.Ctx) error {
	runID, err := uuid.Parse(c.Params("id"))
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid task run ID"))
	}

	service := newTaskRunService()
	run, err := service.GetRun(c.Context(), runID)
	if err != nil {
		return response.Error(c, err)
	}
	if err := ensureRunAccess(c, run); err != nil {
		return response.Error(c, err)
	}
	steps, err := service.ListStepsByRunID(c.Context(), runID)
	if err != nil {
		return response.Error(c, err)
	}
	stepKeysByID := map[uuid.UUID]string{}
	for _, step := range steps {
		stepKeysByID[step.ID] = step.StepKey
	}
	events, err := service.ListEventsByRunID(c.Context(), runID)
	if err != nil {
		return response.Error(c, err)
	}
	items := make([]dto.TaskRunEventResponse, len(events))
	for i, event := range events {
		var stepKey *string
		if event.TaskRunStepID != nil {
			if key, ok := stepKeysByID[*event.TaskRunStepID]; ok {
				stepKey = &key
			}
		}
		items[i] = toTaskRunEventResponse(event, stepKey)
	}
	return response.Success(c, fiber.Map{"events": items})
}

File: backend/src/api/taskrun/handlers.rs
use axum::{
    Json,
    extract::{Path, Query, State},
};
use uuid::Uuid;

use crate::{
    api::{
        auth::{AuthState, AuthenticatedAccount},
        response::SuccessResponse,
    },
    core::{
        auth::AccountId,
        taskrun::{TaskRun, TaskRunFilter},
    },
    error::AppError,
};

use super::{
    dto::{
        TaskRunDetailResponse, TaskRunEventResponse, TaskRunEventsResponse, TaskRunListQuery,
        TaskRunListResponse, TaskRunResponse, step_keys,
    },
    state::TaskRunState,
};

type ApiResult<T> = Result<Json<SuccessResponse<T>>, AppError>;

#[utoipa::path(
    get,
    path = "/task-runs",
    params(TaskRunListQuery),
    responses(
        (status = 200, body = SuccessResponse<TaskRunListResponse>),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "task-runs",
    operation_id = "listTaskRuns"
)]
pub async fn list_task_runs(
    authenticated: AuthenticatedAccount,
    State(task_runs): State<TaskRunState>,
    State(auth): State<AuthState>,
    Query(query): Query<TaskRunListQuery>,
) -> ApiResult<TaskRunListResponse> {
    let account_id = authenticated.into_inner();
    let organization_id = organization_id(&auth, account_id).await;
    let runs = task_runs
        .service()?
        .list_runs(TaskRunFilter {
            organization_id,
            user_id: organization_id.is_none().then_some(account_id.into_inner()),
            task_name: query.task_name.unwrap_or_default(),
            status: query.status.unwrap_or_default(),
            kind: query.kind.unwrap_or_default(),
            limit: query
                .limit
                .as_deref()
                .and_then(|value| value.parse::<i64>().ok())
                .unwrap_or(50),
        })
        .await?
        .into_iter()
        .map(Into::into)
        .collect();
    Ok(Json(SuccessResponse::new(TaskRunListResponse { runs })))
}

#[utoipa::path(
    get,
    path = "/task-runs/{id}",
    params(("id" = String, Path)),
    responses(
        (status = 200, body = SuccessResponse<TaskRunDetailResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "task-runs",
    operation_id = "getTaskRun"
)]
pub async fn get_task_run(
    authenticated: AuthenticatedAccount,
    State(task_runs): State<TaskRunState>,
    State(auth): State<AuthState>,
    Path(id): Path<String>,
) -> ApiResult<TaskRunDetailResponse> {
    let run_id = parse_run_id(&id)?;
    let service = task_runs.service()?;
    let run = service.get_run(run_id).await?;
    ensure_run_access(&auth, authenticated.into_inner(), &run).await?;
    let steps = service.list_steps_by_run_id(run_id).await?;
    let step_keys = step_keys(&steps);
    let events = service
        .list_events_by_run_id(run_id)
        .await?
        .into_iter()
        .map(|event| TaskRunEventResponse::new(event, &step_keys))
        .collect();
    Ok(Json(SuccessResponse::new(TaskRunDetailResponse {
        run: TaskRunResponse::from(run),
        steps: steps.into_iter().map(Into::into).collect(),
        events,
    })))
}

#[utoipa::path(
    get,
    path = "/task-runs/{id}/events",
    params(("id" = String, Path)),
    responses(
        (status = 200, body = SuccessResponse<TaskRunEventsResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "task-runs",
    operation_id = "listTaskRunEvents"
)]
pub async fn list_task_run_events(
    authenticated: AuthenticatedAccount,
    State(task_runs): State<TaskRunState>,
    State(auth): State<AuthState>,
    Path(id): Path<String>,
) -> ApiResult<TaskRunEventsResponse> {
    let run_id = parse_run_id(&id)?;
    let service = task_runs.service()?;
    let run = service.get_run(run_id).await?;
    ensure_run_access(&auth, authenticated.into_inner(), &run).await?;
    let steps = service.list_steps_by_run_id(run_id).await?;
    let step_keys = step_keys(&steps);
    let events = service
        .list_events_by_run_id(run_id)
        .await?
        .into_iter()
        .map(|event| TaskRunEventResponse::new(event, &step_keys))
        .collect();
    Ok(Json(SuccessResponse::new(TaskRunEventsResponse { events })))
}

fn parse_run_id(id: &str) -> Result<Uuid, AppError> {
    Uuid::parse_str(id).map_err(|_| AppError::InvalidInput("Invalid task run ID".to_owned()))
}

async fn organization_id(auth: &AuthState, account_id: AccountId) -> Option<Uuid> {
    auth.service()
        .get_account(account_id)
        .await
        .ok()
        .and_then(|account| account.organization_id)
}

async fn ensure_run_access(
    auth: &AuthState,
    account_id: AccountId,
    run: &TaskRun,
) -> Result<(), AppError> {
    if let Some(organization_id) = organization_id(auth, account_id).await {
        if run.organization_id == Some(organization_id) {
            return Ok(());
        }
        return Err(AppError::NotFound);
    }
    if run.user_id == Some(account_id.into_inner()) {
        Ok(())
    } else {
        Err(AppError::NotFound)
    }
}
